samples/extensions/khr/externalmemory/main.c

```diff
@@ -0,0 +1,746 @@
+// - VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2 required for the previous one
+//   and for querying the device's UUID.
+const char* const required_instance_extensions[] = {
+    VK_KHR_EXTERNAL_MEMORY_CAPABILITIES_EXTENSION_NAME, /*VK_KHR_external_memory_capabilities*/
+    VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2_EXTENSION_NAME /*VK_KHR_get_physical_device_properties2*/
+};
+const size_t required_instance_extensions_count =
+    sizeof(required_instance_extensions) / sizeof(const char*);
+
+// General Vulkan extensions that a device needs to support for exporting
+// memory.
+const char* required_device_extensions[] = {
+    VK_KHR_EXTERNAL_MEMORY_EXTENSION_NAME, /*VK_KHR_external_memory*/
+#ifdef _WIN32
+    VK_KHR_EXTERNAL_MEMORY_WIN32_EXTENSION_NAME /*VK_KHR_external_memory_win32*/
+#else
+    VK_KHR_EXTERNAL_MEMORY_FD_EXTENSION_NAME /*VK_KHR_external_memory_fd*/
+#endif
+};
+const size_t required_device_extensions_count =
+    sizeof(required_device_extensions) / sizeof(const char*);
+
+// Khronos extensions that a device needs to support memory sharing with Vulkan.
+const char* required_khronos_extensions[] = {
+#ifdef _WIN32
+    CL_KHR_EXTERNAL_MEMORY_WIN32_EXTENSION_NAME /*cl_khr_external_memory_win32*/
+#else
+    CL_KHR_EXTERNAL_MEMORY_OPAQUE_FD_EXTENSION_NAME /*cl_khr_external_memory_opaque_fd*/
+#endif
+};
+const size_t required_khronos_extensions_count =
+    sizeof(required_khronos_extensions) / sizeof(const char*);
+
+// Required Vulkan external memory handle.
+const VkExternalMemoryHandleTypeFlagBits vk_external_memory_handle_type =
+#ifdef _WIN32
+    VK_EXTERNAL_MEMORY_HANDLE_TYPE_OPAQUE_WIN32_BIT_KHR;
+#else
+    VK_EXTERNAL_MEMORY_HANDLE_TYPE_OPAQUE_FD_BIT_KHR;
+#endif
+
+// Required OpenCL external memory handle.
+const cl_external_memory_handle_type_khr cl_external_memory_handle_type =
+#ifdef _WIN32
+    CL_EXTERNAL_MEMORY_HANDLE_OPAQUE_WIN32_KHR;
+#else
+    CL_EXTERNAL_MEMORY_HANDLE_OPAQUE_FD_KHR;
+#endif
+
+// Check if a given OpenCL device supports a particular external memory handle
+// type.
+bool cl_check_external_memory_handle_type(
+    const cl_device_id cl_device,
+    cl_external_memory_handle_type_khr external_memory_handle_type)
+{
+    cl_external_memory_handle_type_khr* supported_handle_types = NULL;
+    size_t supported_handle_types_count = 0;
+    cl_int error = CL_SUCCESS;
+
+    OCLERROR_RET(
+        clGetDeviceInfo(cl_device,
+                        CL_DEVICE_EXTERNAL_MEMORY_IMPORT_HANDLE_TYPES_KHR, 0,
+                        NULL, &supported_handle_types_count),
+        error, err);
+    supported_handle_types = (cl_external_memory_handle_type_khr*)malloc(
+        supported_handle_types_count);
+
+    OCLERROR_RET(
+        clGetDeviceInfo(
+            cl_device, CL_DEVICE_EXTERNAL_MEMORY_IMPORT_HANDLE_TYPES_KHR,
+            supported_handle_types_count, supported_handle_types, NULL),
+        error, err);
+    for (size_t i = 0; i < supported_handle_types_count; ++i)
```

Comment:
`supported_handle_types_count` is actually the size in bytes rather than a count, so this loop will read off the end of the allocation.  There needs to be a divide by `cl_external_memory_handle_type_khr` somewhere to compute the count.